frontend/lib/resolution/Resolver.cpp

```diff
@@ -4050,174 +4173,386 @@ void Resolver::exit(const New* node) {
   }
 }
 
-static QualifiedType resolveSerialIterType(Resolver& resolver,
-                                           const AstNode* astForErr,
-                                           const AstNode* iterand) {
-  Context* context = resolver.context;
-  iterand->traverse(resolver);
-  ResolvedExpression& iterandRE = resolver.byPostorder.byAst(iterand);
+static QualifiedType
+getIterKindConstantOrUnknown(Resolver& rv, UniqueString constant) {
+  if (!constant.isEmpty()) {
+    auto ik = EnumType::getIterKindType(rv.context);
+    if (auto m = EnumType::getParamConstantsMapOrNull(rv.context, ik)) {
+      auto it = m->find(constant);
+      if (it != m->end()) return it->second;
+    }
+  }
+  return QualifiedType(QualifiedType::VAR, UnknownType::get(rv.context));
+}
 
-  if (resolver.scopeResolveOnly) {
-    return QualifiedType(QualifiedType::UNKNOWN,
-                         UnknownType::get(context));
+static IterDetails resolveIterDetails(Resolver& rv,
+                                      const AstNode* astForErr,
+                                      const AstNode* iterand,
+                                      const QualifiedType& leaderYieldType,
+                                      int mask) {
+  Context* context = rv.context;
+
+  if (mask == IterDetails::NONE || rv.scopeResolveOnly) {
+    iterand->traverse(rv);
+    return {};
   }
 
-  auto& MSC = iterandRE.mostSpecific();
-  bool isIter = MSC.isEmpty() == false &&
-                MSC.numBest() == 1 &&
-                MSC.only().fn()->untyped()->kind() == uast::Function::Kind::ITER;
+  // Resolve the iterand but suppress errors for now. We'll reissue them
+  // next, possibly suppressing a "NoMatchingCandidates" for the iterand if
+  // our injected call is successful.
```

Comment:
I just needed to resolve the iterand's type fully before I could make any decisions about it. It could be an iterator call. It could be as you say, a type with a standalone iterator but not a serial, in which case we need to inject the arguments and try again.

This is one way to do it, but I agree about the potential negative performance impact. Another option could be to add some sort of `bool doNotEmitCallErrors` flag to the `Resolver`. I tried to get that to work for a bit but I wasn't happy with it.

I will add a future work item about exploring an alternative to `runAndTrackErrors`.